charts/vela-core/templates/defwithtemplate/env.yaml

```diff
@@ -15,18 +15,88 @@ spec:
-        	// +patchKey=name
-        	containers: [{
+        #PatchArgs: {
+        	name:    *"" | string
+        	replace: *false | bool
+        	env: [string]: string
+        	unset: *[] | [...string]
+        }
+        PatchContainer: {
+        	_args: #PatchArgs
+        	name:  string
+        	if _args.name != "" {
+        		name: _args.name
+        	}
+        	if _args.name == "" {
         		name: context.name
-        		// +patchStrategy=retainKeys
-        		env: [
-        			for k, v in parameter.env {
+        	}
+        	_delKeys: {for k in _args.unset {"\(k)": ""}}
+        	_baseContainers: context.output.spec.template.spec.containers
+        	_matchContainers_: [ for _container_ in _baseContainers if _container_.name == name {_container_}]
+        	_baseContainer: *_|_ | {...}
+        	if len(_matchContainers_) == 0 {
+        		err: "container \(name) not found"
+        	}
+        	if len(_matchContainers_) > 0 {
+        		_baseContainer: _matchContainers_[0]
+        		_baseEnv:       _baseContainer.env
+        		if _baseEnv == _|_ {
+        			// +patchStrategy=replace
+        			env: [ for k, v in _args.env if _delKeys[k] == _|_ {
+        				name:  k
+        				value: v
+        			}]
+        		}
+        		if _baseEnv != _|_ {
+        			_baseEnvMap: {for envVar in _baseEnv {"\(envVar.name)": envVar.value}}
+        			// +patchStrategy=replace
+        			env: [ for envVar in _baseEnv if _delKeys[envVar.name] == _|_ && !_args.replace {
+        				name: envVar.name
+        				if _args.env[envVar.name] != _|_ {
+        					value: _args.env[envVar.name]
+        				}
+        				if _args.env[envVar.name] == _|_ {
+        					value: envVar.value
+        				}
+        			}] + [ for k, v in _args.env if _delKeys[k] == _|_ && (_args.replace || _baseEnvMap[k] == _|_) {
         				name:  k
         				value: v
-        			},
-        		]
-        	}]
+        			}]
+        		}
+        	}
+        }
+        patch: spec: template: spec: {
+        	if parameter.containers == _|_ {
+        		// +patchKey=name
+        		containers: [{
+        			PatchContainer & {_args: {
+        				if parameter.name == "" {
+        					name: context.name
+        				}
+        				if parameter.name != "" {
+        					name: parameter.name
+        				}
+        				replace: parameter.replace
+        				env:     parameter.env
+        				unset:   parameter.unset
+        			}}
+        		}]
+        	}
+        	if parameter.containers != _|_ {
+        		// +patchKey=name
+        		containers: [ for c in parameter.containers {
+        			if c.name == "" {
+        				err: "name must be set for containers"
+        			}
+        			if c.name != "" {
+        				PatchContainer & {_args: c}
+        			}
+        		}]
+        	}
+        }
+        parameter: {
+        	#PatchArgs
+        	containers?: [...#PatchArgs]
         }
-        parameter: env: [string]: string
```

Comment:
It is best to keep the previous definition and make it deprecated and then create a new one, otherwise, it will affect the users who use the old definition.

Reply:
Ah, I see, the previous format is still supported.

Reply:
Right, I was also thinking of making this trait `container-env`, which is more accurate. But it might be better to keep them converge together.